File: ios/SalmonTrollingAssistant/SalmonTrollingAssistant/Domain/Services/AuthenticationServiceImpl.swift
import Foundation
import Combine
import KeychainAccess

class AuthenticationServiceImpl: AuthenticationService {
    private let keychain = Keychain(service: "com.example.SalmonTrollingAssistant")
    private let userDefaults = UserDefaults.standard
    
    private let userProfileKey = "userProfile"
    private let authTokenKey = "authToken"
    
    let currentUser = CurrentValueSubject<UserProfile?, Never>(nil)
    
    var isSignedIn: AnyPublisher<Bool, Never> {
        return currentUser
            .map { $0 != nil }
            .eraseToAnyPublisher()
    }
    
    var isAnonymousUser: AnyPublisher<Bool, Never> {
        return currentUser
            .map { $0?.isAnonymous ?? false }
            .eraseToAnyPublisher()
    }
    
    init() {
        // Load the current user on initialization
        if let userData = userDefaults.data(forKey: userProfileKey),
           let userProfile = try? JSONDecoder().decode(UserProfile.self, from: userData) {
            currentUser.send(userProfile)
        }
    }
    
    func createAccount(credentials: AuthCredentials, name: String?) async -> AuthResult {
        // In a real app, this would make an API call to create an account
        // For now, we'll simulate account creation locally
        
        // Check if email already exists
        if emailExists(credentials.email) {
            return AuthResult(
                success: false,
                error: "Email already in use"
            )
        }
        
        let userId = UUID().uuidString
        let token = UUID().uuidString
        
        // Create and save the user profile
        let userProfile = UserProfile(
            id: userId,
            name: name,
            email: credentials.email,
            isAnonymous: false
        )
        
        saveCurrentUser(userProfile)
        saveCredentials(credentials)
        saveAuthToken(token)
        
        return AuthResult(
            success: true,
            userId: userId,
            token: token
        )
    }
    
    func signIn(credentials: AuthCredentials) async -> AuthResult {
        // In a real app, this would make an API call to authenticate
        // For now, we'll simulate authentication locally
        
        guard let savedPassword = getSavedPassword(for: credentials.email),
              savedPassword == credentials.password else {
            return AuthResult(
                success: false,
                error: "Invalid email or password"
            )
        }
        
        // Load or create user profile
        let currentUserValue = currentUser.value
        let userId = currentUserValue?.id ?? UUID().uuidString
        let token = UUID().uuidString
        
        let userProfile: UserProfile
        if var existingUser = currentUserValue {
            existingUser.email = credentials.email
            existingUser.isAnonymous = false
            userProfile = existingUser
        } else {
            userProfile = UserProfile(
                id: userId,
                email: credentials.email,
                isAnonymous: false
            )
        }
        
        saveCurrentUser(userProfile)
        saveAuthToken(token)
        
        return AuthResult(
            success: true,
            userId: userId,
            token: token
        )
    }
    
    func signOut() async -> Bool {
        saveAuthToken(nil)
        saveCurrentUser(nil)
        return true
    }
    
    func signInAnonymously() async -> AuthResult {
        let userId = UUID().uuidString
        let token = UUID().uuidString
        
        let userProfile = UserProfile(
            id: userId,
            isAnonymous: true
        )
        
        saveCurrentUser(userProfile)
        saveAuthToken(token)
        
        return AuthResult(
            success: true,
            userId: userId,
            token: token
        )
    }
    
    func convertAnonymousAccount(credentials: AuthCredentials, name: String?) async -> AuthResult {
        guard var currentUserValue = currentUser.value else {
            return AuthResult(
                success: false,
                error: "No anonymous user to convert"
            )
        }
        
        if !currentUserValue.isAnonymous {
            return AuthResult(
                success: false,
                error: "Current user is not anonymous"
            )
        }
        
        // Check if email already exists
        if emailExists(credentials.email) {
            return AuthResult(
                success: false,
                error: "Email already in use"
            )
        }
        
        let token = UUID().uuidString
        
        currentUserValue.name = name
        currentUserValue.email = credentials.email
        currentUserValue.isAnonymous = false
        
        saveCurrentUser(currentUserValue)
        saveCredentials(credentials)
        saveAuthToken(token)
        
        return AuthResult(
            success: true,
            userId: currentUserValue.id,
            token: token
        )
    }
    
    // MARK: - Helper Methods
    
    private func saveCurrentUser(_ userProfile: UserProfile?) {
        if let userProfile = userProfile,
           let userData = try? JSONEncoder().encode(userProfile) {
            userDefaults.set(userData, forKey: userProfileKey)
        } else {
            userDefaults.removeObject(forKey: userProfileKey)
        }
        
        currentUser.send(userProfile)
    }
    
    private func emailExists(_ email: String) -> Bool {
        return (try? keychain.get("email_\(email)")) != nil
    }
    
    private func saveCredentials(_ credentials: AuthCredentials) {
        try? keychain.set(credentials.password, key: "email_\(credentials.email)")
    }
    
    private func getSavedPassword(for email: String) -> String? {
        return try? keychain.get("email_\(email)")
    }
    
    private func saveAuthToken(_ token: String?) {
        if let token = token {
            try? keychain.set(token, key: authTokenKey)
        } else {
            try? keychain.remove(authTokenKey)
        }
    }
    
    private func getAuthToken() -> String? {
        return try? keychain.get(authTokenKey)
    }
}

File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/data/service/AuthenticationServiceImpl.kt
package com.example.salmontrollingassistant.data.service

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.example.salmontrollingassistant.domain.model.AuthCredentials
import com.example.salmontrollingassistant.domain.model.AuthResult
import com.example.salmontrollingassistant.domain.model.UserProfile
import com.example.salmontrollingassistant.domain.service.AuthenticationService
import com.squareup.moshi.Moshi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.map
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthenticationServiceImpl @Inject constructor(
    private val context: Context,
    private val moshi: Moshi
) : AuthenticationService {
    
    private val masterKey = MasterKey.Builder(context)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()
    
    private val sharedPreferences: SharedPreferences = EncryptedSharedPreferences.create(
        context,
        "auth_prefs",
        masterKey,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )
    
    private val userProfileAdapter = moshi.adapter(UserProfile::class.java)
    
    private val currentUserFlow = MutableStateFlow<UserProfile?>(loadCurrentUser())
    
    private fun loadCurrentUser(): UserProfile? {
        val userJson = sharedPreferences.getString(KEY_USER_PROFILE, null)
        return userJson?.let { userProfileAdapter.fromJson(it) }
    }
    
    private fun saveCurrentUser(userProfile: UserProfile?) {
        if (userProfile == null) {
            sharedPreferences.edit().remove(KEY_USER_PROFILE).apply()
        } else {
            val userJson = userProfileAdapter.toJson(userProfile)
            sharedPreferences.edit().putString(KEY_USER_PROFILE, userJson).apply()
        }
        currentUserFlow.value = userProfile
    }
    
    override suspend fun createAccount(credentials: AuthCredentials, name: String?): AuthResult {
        // In a real app, this would make an API call to create an account
        // For now, we'll simulate account creation locally
        
        // Check if email already exists
        if (emailExists(credentials.email)) {
            return AuthResult(
                success = false,
                error = "Email already in use"
            )
        }
        
        val userId = UUID.randomUUID().toString()
        val token = UUID.randomUUID().toString()
        
        // Create and save the user profile
        val userProfile = UserProfile(
            id = userId,
            name = name,
            email = credentials.email,
            isAnonymous = false
        )
        
        saveCurrentUser(userProfile)
        saveCredentials(credentials.email, credentials.password)
        saveAuthToken(token)
        
        return AuthResult(
            success = true,
            userId = userId,
            token = token
        )
    }
    
    override suspend fun signIn(credentials: AuthCredentials): AuthResult {
        // In a real app, this would make an API call to authenticate
        // For now, we'll simulate authentication locally
        
        val savedPassword = getSavedPassword(credentials.email)
        if (savedPassword == null || savedPassword != credentials.password) {
            return AuthResult(
                success = false,
                error = "Invalid email or password"
            )
        }
        
        // Load or create user profile
        val currentUser = loadCurrentUser()
        val userId = currentUser?.id ?: UUID.randomUUID().toString()
        val token = UUID.randomUUID().toString()
        
        val userProfile = currentUser?.copy(
            email = credentials.email,
            isAnonymous = false
        ) ?: UserProfile(
            id = userId,
            email = credentials.email,
            isAnonymous = false
        )
        
        saveCurrentUser(userProfile)
        saveAuthToken(token)
        
        return AuthResult(
            success = true,
            userId = userId,
            token = token
        )
    }
    
    override suspend fun signOut(): Boolean {
        saveAuthToken(null)
        saveCurrentUser(null)
        return true
    }
    
    override suspend fun signInAnonymously(): AuthResult {
        val userId = UUID.randomUUID().toString()
        val token = UUID.randomUUID().toString()
        
        val userProfile = UserProfile(
            id = userId,
            isAnonymous = true
        )
        
        saveCurrentUser(userProfile)
        saveAuthToken(token)
        
        return AuthResult(
            success = true,
            userId = userId,
            token = token
        )
    }
    
    override suspend fun convertAnonymousAccount(credentials: AuthCredentials, name: String?): AuthResult {
        val currentUser = loadCurrentUser() ?: return AuthResult(
            success = false,
            error = "No anonymous user to convert"
        )
        
        if (!currentUser.isAnonymous) {
            return AuthResult(
                success = false,
                error = "Current user is not anonymous"
            )
        }
        
        // Check if email already exists
        if (emailExists(credentials.email)) {
            return AuthResult(
                success = false,
                error = "Email already in use"
            )
        }
        
        val token = UUID.randomUUID().toString()
        
        val updatedUser = currentUser.copy(
            name = name,
            email = credentials.email,
            isAnonymous = false
        )
        
        saveCurrentUser(updatedUser)
        saveCredentials(credentials.email, credentials.password)
        saveAuthToken(token)
        
        return AuthResult(
            success = true,
            userId = currentUser.id,
            token = token
        )
    }
    
    override fun getCurrentUser(): Flow<UserProfile?> {
        return currentUserFlow
    }
    
    override fun isSignedIn(): Flow<Boolean> {
        return currentUserFlow.map { it != null }
    }
    
    override fun isAnonymousUser(): Flow<Boolean> {
        return currentUserFlow.map { it?.isAnonymous ?: false }
    }
    
    // Helper methods for credential management
    private fun emailExists(email: String): Boolean {
        return sharedPreferences.contains("email_$email")
    }
    
    private fun saveCredentials(email: String, password: String) {
        sharedPreferences.edit()
            .putString("email_$email", password)
            .apply()
    }
    
    private fun getSavedPassword(email: String): String? {
        return sharedPreferences.getString("email_$email", null)
    }
    
    private fun saveAuthToken(token: String?) {
        if (token == null) {
            sharedPreferences.edit().remove(KEY_AUTH_TOKEN).apply()
        } else {
            sharedPreferences.edit().putString(KEY_AUTH_TOKEN, token).apply()
        }
    }
    
    private fun getAuthToken(): String? {
        return sharedPreferences.getString(KEY_AUTH_TOKEN, null)
    }
    
    companion object {
        private const val KEY_USER_PROFILE = "user_profile"
        private const val KEY_AUTH_TOKEN = "auth_token"
    }
}